csrc/executor.cpp

```diff
@@ -1315,35 +1314,13 @@ KernelArgumentHolder FusionExecutor::inferOutputSizes(
           fusion->outputs()[out_i]->isA<TensorView>(),
           "Cannot allocate outputs that are not tensors.");
       auto output_tv = fusion->outputs()[out_i]->as<TensorView>();
-
-      auto alias_it = std::find_if(
-          output_to_input_aliases.begin(),
-          output_to_input_aliases.end(),
-          [&](const auto& output_to_input) {
-            return output_to_input.first == (int)out_i;
-          });
-
-      if (alias_it != output_to_input_aliases.end()) {
-        // When aliasing output to an input, we do not need to allocate a new
-        // output but still need to push an entry.
-        ret.push(PolymorphicValue(0L));
-      } else {
-        const auto& [sizes, strides] = inferShapeOfOutput(output_tv, expr_eval);
-        const auto dtype = (output_tv->dtype() == DataType::Index)
-            ? data_type_to_aten(arg_index_type)
-            : data_type_to_aten(output_tv->dtype());
-        ret.pushTensorProxy(sizes, strides, dtype);
-      }
+      const auto& [sizes, strides] = inferShapeOfOutput(output_tv, expr_eval);
```

Comment:
The logic here looks fine.

Looking at allocOutputs part here, everything seems fine as well: https://github.com/NVIDIA/Fuser/blob/main/csrc/executor.cpp#L916

We are also skipping it here:
https://github.com/NVIDIA/Fuser/blob/main/csrc/kernel_cache.cpp#L198C1-L206

Since we are not only pushing TensorProxy here, maybe it's perfectly fine to push trivial forward as-is.

Can we double check that part with some simple examples?

Reply:
> Can we double check that part with some simple examples?

Double check a trivially forwarded input stays an alias? Sure, I added more checks in `FusionTrivialInputForwarding*_CUDA`.

Reply:
I think more importantly, double checking that we are NOT allocating any memory for the trivial forwarded input and throwing it away later...
Having said that, I don't know what would be a reliable way to check that automatically though.

Reply:
Phew, it took me a while to confirm this, but I managed to verify the number of calls to `cudaMalloc` equals the number of inputs. 

```
me @ swagtron : dev | /opt/pytorch/nvfuser (alias*)
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 nsys nvprof bin/nvfuser_tests --gtest_filter=*TrivialInputForwarding2_CUDA 2> /dev/null | grep cudaMalloc
      1.2            94359          1    94359.0    94359.0     94359     94359          0.0  cudaMalloc
me @ swagtron : dev | /opt/pytorch/nvfuser (alias*)
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 nsys nvprof bin/nvfuser_tests --gtest_filter=*TrivialInputForwarding_CUDA 2> /dev/null | grep cudaMalloc
      1.2           100611          2    50305.5    50305.5      6102     94509      62513.2  cudaMalloc
```

Two caveats though: 
1. I commented out `testValidate` which allocates more tensors for computing atol and rtol. 
2. Surprisingly, `at::empty` triggers a `cudaMalloc`! So I patched the following. This patch may be the right thing to do anyway. Trivial input forwarding is a special case of the aliasing I'm trying to support.

```
diff --git a/csrc/executor.cpp b/csrc/executor.cpp
index e1ad311..f9ebe93 100644
--- a/csrc/executor.cpp
+++ b/csrc/executor.cpp
@@ -943,6 +943,15 @@ std::vector<at::Tensor> allocOutputs(
           "alias io only supports tensor");
       outputs.emplace_back(*inputs[aliased_input_index]);
     } else if (kernel->outputs().at(output_idx)->isFusionInput()) {
+#if 1
+      const int input_idx = std::find(
+                                kernel->inputs().begin(),
+                                kernel->inputs().end(),
+                                kernel->outputs().at(output_idx)) -
+          kernel->inputs().begin();
+      NVF_ERROR(input_idx < (int)kernel->inputs().size());
+      outputs.emplace_back(*inputs[input_idx]);
+#else
       // pushing empty tensor for trivial forwarding. Since we handle this in
       // integration, see step 1 - note [trivial forwarding]
       auto alloc_dom =
@@ -954,6 +963,7 @@ std::vector<at::Tensor> allocOutputs(
           at::TensorOptions().dtype(at::kFloat).device(device);
       outputs.emplace_back(
           at::empty(std::vector<int64_t>(alloc_dom.size(), 0), tensor_options));
+#endif
     } else {
       auto alloc_tensor = at::native::empty_strided_cuda(
           buf_info.sizes,
```

Reply:
Thanks for double checking this~~

That's indeed a surprise, I thought `at::empty` with a zero sized-tensor should have got a short-cut to not trigger any allocation call there. :laughing: 

Also agree that trivial forwarding is just a special case of aliasing. Lots of clean up and code merge to be done :+1: